lib/three_scale/middleware/presigned_downloads.rb

```diff
@@ -0,0 +1,88 @@
+# frozen_string_literal: true
+
+module ThreeScale
+  module Middleware
+    class PresignedDownloads
+      RE_SEPARATOR = ::File::SEPARATOR == "/" ? "/" : "[#{Regexp.escape(::File::SEPARATOR)}/]"
+
+      delegate *%i[requires_signing? good_signature? signature_at to_time], to: :class, private: true
+
+      class << self
+        def verifier
+          Rails.application.message_verifier("file downloads")
+        end
+
+        def requires_signing?(path)
+          # for the used separator see https://github.com/rack/rack/pull/2257
+          # question mark on first separator is paranoia of getting a relative path somehow
+          invoices_re = %r{#{RE_SEPARATOR}?system#{RE_SEPARATOR}.*#{RE_SEPARATOR}invoices#{RE_SEPARATOR}}
+          path.start_with? invoices_re
+        end
+
+        def sign_path_if_needed(path, expires_in = 3600)
+          requires_signing?(path) ? sign(path, expires_in) : path
+        end
+
+        def sign(path, expires_in = 3600)
+          sign_at(path, Time.now.utc.round + expires_in)
+        end
+
+        def sign_at(path, expires_at)
+          expires_timestamp = to_timestamp(expires_at)
+          signature = signature_at(path, expires_at)
+          params = { "3scale-Expires" => expires_timestamp, "3scale-Signature" => signature }.to_param
+
+          "#{path}?#{params}"
+        end
+
+        def signature_at(path, expires_at)
+          signing_options = { purpose: :download, expires_at: }
+          verifier.generate(path.b, **signing_options).split("--").last
+        end
+
+        def to_timestamp(time)
+          time.iso8601.gsub(/[-:]/, "")
+        end
+
+        def to_time(timestamp)
+          Time.strptime(timestamp, "%Y%m%dT%H%M%S%Z") rescue Time.new(1970)
```

Comment:
Better use UTC consistently in all calls to Time I think:

```suggestion
          Time.strptime(timestamp, "%Y%m%dT%H%M%S%Z") rescue Time.new.utc(1970)
```

I'm not sure there's actually a difference here, but Rubocop doesn't like it.

Reply:
It doesn't matter. It is just time in the past. I'm not in favor of changing code to make a linter happy when it is not adding value. But this would be nicer to be some constant so I'll change it like that.

Reply:
8032a388a